### ISSUE TITLE
Collation dependent issue in TiDB on `select * from t1 where v like` 

### ISSUE THREAD

**Issue by espresso98** [pingcap/tidb#31022]
## Bug Report

I enabled the collation framework but the result is different as in MySQL.

### 1. Minimal reproduce step 
```sql
drop table if exists t1;
create table t1 (v varchar(254), index (v)) charset=utf8mb4 collate=utf8mb4_bin;
insert into t1 values ("This is a test "),(" This is a test "),("This is a test"),("This is a test");
select * from t1 where v like 'This is a test';
select * from t1 where v='This is a test';
```

### 2. What did you expect to see? 
In MySQL5.7 and 8.0,
```sql
mysql> select * from t1 where v like 'This is a test';
+----------------+
| v              |
+----------------+
| This is a test |
| This is a test |
+----------------+
2 rows in set (0.00 sec)

mysql> select * from t1 where v='This is a test';
+-----------------+
| v               |
+-----------------+
| This is a test  |
| This is a test  |
| This is a test  |
+-----------------+
3 rows in set (0.01 sec)
```

### 3. What did you see instead 
In TiDB,
```sql
mysql> select * from t1 where v like 'This is a test';
+-----------------+
| v               |
+-----------------+
| This is a test  |
| This is a test  |
| This is a test  |
+-----------------+
3 rows in set (0.00 sec)

mysql> select * from t1 where v='This is a test';
+-----------------+
| v               |
+-----------------+
| This is a test  |
| This is a test  |
| This is a test  |
+-----------------+
3 rows in set (0.00 sec)

```

### 4. What is your TiDB version? 
```sql
tidb_version(): Release Version: v5.4.0-alpha-459-g87ab28ebb-dirty
Edition: Community
Git Commit Hash: 87ab28ebb2ac87b6e8d7c252187338c517598318
Git Branch: master
UTC Build Time: 2021-12-20 19:44:29
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)
```

**Comment by xiongjiwei** [comment]
```sql
mysql> desc select * from t1 where v like 'This is a test';
+------------------------+---------+-----------+----------------------+---------------------------------------------------------------------------+
| id                     | estRows | task      | access object        | operator info                                                             |
+------------------------+---------+-----------+----------------------+---------------------------------------------------------------------------+
| IndexReader_6          | 0.00    | root      |                      | index:IndexRangeScan_5                                                    |
| └─IndexRangeScan_5     | 0.00    | cop[tikv] | table:t1, index:v(v) | range:["This is a test","This is a test"], keep order:false, stats:pseudo |
+------------------------+---------+-----------+----------------------+---------------------------------------------------------------------------+
2 rows in set (0.00 sec)
```

`like` in this test has been rewritten to `eq`, they are different from the tailing space.

**Comment by Reminiscent** [comment]
In the lastest master, we only get two rows.
```
mysql> drop table if exists t1;
Query OK, 0 rows affected (0.03 sec)

mysql> create table t1 (v varchar(254), index (v)) charset=utf8mb4 collate=utf8mb4_bin;
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t1 values ("This is a test "),(" This is a test "),("This is a test"),("This is a test");
Query OK, 4 rows affected (0.02 sec)
Records: 4  Duplicates: 0  Warnings: 0

mysql> select * from t1 where v like 'This is a test';
+----------------+
| v              |
+----------------+
| This is a test |
| This is a test |
+----------------+
2 rows in set (0.02 sec)

mysql> select * from t1 where v='This is a test';
+----------------+
| v              |
+----------------+
| This is a test |
| This is a test |
+----------------+
2 rows in set (0.00 sec)
```

**Comment by Reminiscent** [comment]
The root cause is that the `like` condition is used to construct the range. But range does not filter the cases with trailing spaces, while the `like` condition can filter. One way to ensure that there is no regression is to use the `like` condition to construct a range while preserving the `like` expressions as a selection. But in the current implementation, it's not easy to do that.

**Comment by xhebox** [comment]
Closed by #31278
